app/src/main/java/com/duckduckgo/app/browser/BrowserTabFragment.kt

```diff
@@ -535,35 +536,91 @@ class BrowserTabFragment :
             credentials: List<LoginCredentials>,
             triggerType: LoginTriggerType,
         ) {
-            showAutofillDialogChooseCredentials(originalUrl, credentials, triggerType)
+            withContext(dispatchers.main()) {
+                showAutofillDialogChooseCredentials(originalUrl, credentials, triggerType)
+            }
+        }
+
+        override suspend fun onGeneratedPasswordAvailableToUse(
+            originalUrl: String,
+            username: String?,
+            generatedPassword: String,
+        ) {
+            // small delay added to let keyboard disappear if it was present; helps avoid jarring transition
+            delay(100)
+
+            withContext(dispatchers.main()) {
+                showUserAutoGeneratedPasswordDialog(originalUrl, username, generatedPassword)
+            }
         }
 
         override fun noCredentialsAvailable(originalUrl: String) {
             viewModel.returnNoCredentialsWithPage(originalUrl)
         }
 
+        override fun onCredentialsSaved(savedCredentials: LoginCredentials) {
+            viewModel.onShowUserCredentialsSaved(savedCredentials)
+        }
+
         override suspend fun onCredentialsAvailableToSave(
             currentUrl: String,
             credentials: LoginCredentials,
         ) {
-            val username = credentials.username
-            val password = credentials.password
+            withContext(dispatchers.main()) {
+                val username = credentials.username
+                val password = credentials.password
 
-            if (username == null && password == null) {
-                Timber.w("Not saving credentials with null username and password")
-                return
+                if (username == null && password == null) {
+                    Timber.w("Not saving credentials with null username and password")
+                    return@withContext
+                }
+
+                val matchType = existingCredentialMatchDetector.determine(currentUrl, username, password)
+                Timber.v("MatchType is %s", matchType.javaClass.simpleName)
+
+                // we need this delay to ensure web navigation / form submission events aren't blocked
+                delay(100)
```

Comment:
Related to the above, this will block main(), is that intended? This used to block default so double checking this is correct.

Reply:
Yeah it's correct. Even though the coroutine is suspended the main thread isn't blocked, so the `delay` is safe to use on the main dispatcher